kotlin/goodmetrics/src/test/kotlin/goodmetrics/pipeline/AggregatorTest.kt

```diff
@@ -112,4 +129,241 @@ internal class AggregatorTest {
+        assertEquals(160, e.positiveBuckets.size, "bucket count should be sensible after rescale")
+        assertEquals(1, e.positiveBuckets[e.positiveBuckets.size - 1], "last index has a new count")
+        assertEquals(3018, e.bucketStartOffset(), "bucket start offset changes because we scaled and rotated")
+
+        //
+        // -------- Skip several zoom scale steps in a single accumulate --------
+        //
+
+        val recursiveScaleStartCount = e.count()
+        assertEquals(2199023255551.996, lowerBoundary(2, 164u), "this value gets us down into scale 2")
+        assertEqualsEpsilon(4.0, lowerBoundary(2, 8u), "this value gets us down into scale 2")
+        assertEqualsEpsilon(4.757, lowerBoundary(2, 9u), "this value gets us down into scale 2")
+        // pin the bucket's low value, at scale 2's index 8. It's not in scale 2 yet though!
+        e.accumulate(4.25)
+        // now blow the range wide, way past scale 7, resulting in a recursive scale down from 7 to precision 2.
+        e.accumulate(2_199_023_255_552.0)
+        assertEquals(2, e.scale(), "this value range should force scale range 2")
+        assertEquals(8, e.bucketStartOffset(), "bucket start offset should match the first element, since we rotated and grew out to the larger value")
+        assertEquals(1, e.positiveBuckets[8 - 8], "this is the 4.0 bucket, and 4.25 should go in it.")
+        assertEquals(1, e.positiveBuckets[164 - 8], "this is the bucket for the big numer.")
+        assertEquals(recursiveScaleStartCount + 2, e.count(), "2 more reports were made. The histogram maintains every count across rescaling, even recursive rescaling")
+    }
+
+    // Helps us look for random index crashes
+    @Test
+    fun testExponentialHistogramFuzz() {
+        val start = System.nanoTime()
+        while ((System.nanoTime() - start).milliseconds < 50.milliseconds) {
+            val e = Aggregation.ExponentialHistogram(8u)
+            val innerStart = System.nanoTime()
+            while ((System.nanoTime() - innerStart).milliseconds < 1.milliseconds) {
+                e.accumulate(1_000_000_000_000_000.0 * Random.nextDouble())
+            }
+        }
+    }
+
+    // Helps us look for random index crashes
+    @Test
+    fun testExponentialHistogramFuzzNegative() {
+        val start = System.nanoTime()
+        while ((System.nanoTime() - start).milliseconds < 50.milliseconds) {
+            val e = Aggregation.ExponentialHistogram(8u)
+            val innerStart = System.nanoTime()
+            while ((System.nanoTime() - innerStart).milliseconds < 1.milliseconds) {
+                e.accumulate(-1_000_000_000_000_000.0 * Random.nextDouble())
+            }
+        }
+    }
+
+    // Verify multi-threaded access is indeed safe
+    @Test
+    fun testExponentialHistogramMultiThreadedAccess() = runBlocking {
```

Comment:
This runs the test on 1 thread with 2 coroutines. This does not exercise multi-threaded access. I'd recommend using simple thread spawn apis and have each of 8 threads accumulate 10000 times each with no delays or sleeps. Join the 8 threads then assert that the accumulation is as you expect it to be.

Reply:
Ah that sounds good, I'll try it with that, thank you

Reply:
https://github.com/kvc0/goodmetrics_kotlin/pull/38/commits/7cff3e3435f844d889e477b5b1558226d636e265